sdk/identity/identity/src/msal/msalClient.ts

```diff
@@ -0,0 +1,260 @@
+ * @param clientId - The client ID of the application.
+ * @param  tenantId - The tenant ID of the Azure Active Directory.
+ * @param  createMsalClientOptions - Optional. Additional options for creating the MSAL client.
+ * @returns  The MSAL configuration object.
+ */
+export function generateMsalConfiguration(
+  clientId: string,
+  tenantId: string,
+  createMsalClientOptions: CreateMsalClientOptions = {},
+): msal.Configuration {
+  const resolvedTenant = resolveTenantId(msalLogger, tenantId, clientId);
+
+  // TODO: move and reuse getIdentityClientAuthorityHost
+  const authority = getAuthority(
+    resolvedTenant,
+    createMsalClientOptions.authorityHost ?? process.env.AZURE_AUTHORITY_HOST,
+  );
+
+  const httpClient = new IdentityClient({
+    ...createMsalClientOptions.tokenCredentialOptions,
+    authorityHost: authority,
+    loggingOptions: createMsalClientOptions.loggingOptions,
+  });
+
+  const msalConfig: msal.Configuration = {
+    auth: {
+      clientId,
+      authority,
+      knownAuthorities: getKnownAuthorities(
+        resolvedTenant,
+        authority,
+        createMsalClientOptions.disableInstanceDiscovery,
+      ),
+    },
+    system: {
+      networkClient: httpClient,
+      loggerOptions: {
+        loggerCallback: defaultLoggerCallback(createMsalClientOptions.logger ?? msalLogger),
+        logLevel: getMSALLogLevel(getLogLevel()),
+        piiLoggingEnabled: createMsalClientOptions.loggingOptions?.enableUnsafeSupportLogging,
+      },
+    },
+  };
+  return msalConfig;
+}
+
+/**
+ * Represents the state necessary for the MSAL (Microsoft Authentication Library) client to operate.
+ * This includes the MSAL configuration, cached account information, Azure region, and a flag to disable automatic authentication.
+ *
+ * @internal
+ */
+interface MsalClientState {
+  /** The configuration for the MSAL client. */
+  msalConfig: msal.Configuration;
+
+  /** The cached account information, or null if no account information is cached. */
+  cachedAccount: msal.AccountInfo | null;
+}
+
+/**
+ * Creates an instance of the MSAL (Microsoft Authentication Library) client.
+ *
+ * @param clientId - The client ID of the application.
+ * @param tenantId - The tenant ID of the Azure Active Directory.
+ * @param createMsalClientOptions - Optional. Additional options for creating the MSAL client.
+ * @returns An instance of the MSAL client.
+ *
+ * @public
+ */
+export function createMsalClient(
+  clientId: string,
+  tenantId: string,
+  createMsalClientOptions: CreateMsalClientOptions = {},
+): MsalClient {
+  const state: MsalClientState = {
+    msalConfig: generateMsalConfiguration(clientId, tenantId, createMsalClientOptions),
+    cachedAccount: createMsalClientOptions.authenticationRecord
+      ? publicToMsal(createMsalClientOptions.authenticationRecord)
+      : null,
+  };
+
+  // configure persistence
+  // configure broker
+
+  let confidentialApp: msal.ConfidentialClientApplication | undefined = undefined;
```

Comment:
My understanding is that each credential will either use a confidential or a public application, but never both. If you look at [msalNodeCommon](https://github.com/Azure/azure-sdk-for-js/blob/e01e73ae96e0b9cd0033c9ae61ddee36aba48bda/sdk/identity/identity/src/msal/nodeFlows/msalNodeCommon.ts#L450-L455) there's a lot of logic that likely is only necessary because we tried to centralize all the code and handle every credential. 

Now, each function will get an MSAL application either via getConfidentialApp or getPublicApp depending on the needs, and should avoid this logic that tries to get things from both applications